test/xbgpu/test_engine.py

```diff
@@ -1452,3 +1555,153 @@ def get_stream_status(stream_name: str) -> bool:
 
             await client.request("capture-start", output.name)
             assert get_stream_status(output.name) is True, f"Stream {output.name} is still disabled"
+
+    @pytest.mark.parametrize(
+        "n_ants, n_channels, n_jones_per_batch, heap_accumulation_threshold",
+        [(4, 1024, 262144, [20, 20])],
+    )
+    @pytest.mark.parametrize("n_x_streams_to_stop, n_b_streams_to_stop", [(2, None), (1, 1), (None, 2)])
+    async def test_capture_stop_some_streams(
+        self,
+        monkeypatch: pytest.MonkeyPatch,
+        mock_recv_streams: list[spead2.InprocQueue],
+        mock_send_stream: list[spead2.InprocQueue],
+        client: aiokatcp.Client,
+        n_ants: int,
+        n_channels_per_substream: int,
+        frequency: int,
+        n_samples_between_spectra: int,
+        n_spectra_per_heap: int,
+        heap_accumulation_threshold: list[int],
+        corrprod_outputs: list[XOutput],
+        beam_outputs: list[BOutput],
+        n_x_streams_to_stop: int,
+        n_b_streams_to_stop: int,
+        xbengine: XBEngine,
+    ) -> None:
+        """Test capture-stop request on X- and B-engine data streams.
+
+        Issue a `?capture-stop` request at some point into data processing and
+        check the corresponding streams only have partial data transmission.
+        Also ensure that data is completely received for data streams that did
+        not receive a `?capture-stop` request.
+
+        This test is carried out in combinations to fully exercise stop logic in
+        both X- and BPipelines.
+        """
+        rng = np.random.default_rng(seed=1)
+        # NOTE: This value is arbitrarily chosen to have a few whole accumulations
+        # for each corrprod_output. This allows us to issue a capture-stop request
+        # partway in its processing.
+        n_total_accumulations = 5
+        n_batches = heap_accumulation_threshold[0] * n_total_accumulations
+        corrprod_capture_stop_accum_indices: list[int | None] = [None] * len(corrprod_outputs)
+        if n_x_streams_to_stop is not None:
```

Comment:
Wouldn't it be simpler to just 0 instead of None, and then just make the code unconditional?

Reply:
Fair enough. I updated the zeroes to `None`s in a naive attempt to be consistent with the other logical tweak you suggested.

Reply:
The difference is that in the other case zero didn't actually mean zero (as in stop after zero chunks) - it triggered a behaviour qualitatively different from other integers. Here zero would really mean zero streams.